lib/flash.js

```diff
@@ -0,0 +1,184 @@
+function factoryPartition(mac1, mac2) {
+  return concatToBuffer(
+    [0x20, 0x76, 0x03, 0x01],
+    mac1,
+    Array(30).fill(0xff),
+    mac2
+  );
+}
+
+function transaction(usbConnection, write, read, status_poll, wren, next) {
+  read = read || 0;
+  status_poll = status_poll || false;
+  wren = wren || false;
+
+  if (write.length > 500 || read >= Math.pow(2, 24)) {
+    return next(new Error('Transaction too large'));
+  }
+  var flags = (status_poll ? 1 : 0) | ((wren ? 1 : 0) << 1);
+  var hdr = [(read >> 0) & 0xff, (read >> 8) & 0xff, (read >> 16) & 0xff, flags];
+
+  var data = concatToBuffer(hdr, write);
+
+  usbConnection.epOut.transfer(data, function(err) {
+    if (err) {
+      return next(err);
+    }
+    if (read > 0) {
+      return usbConnection.epIn.transfer(read, next);
+    }
+    next();
+  });
+}
+
+function readChipId(usbConnection, next) {
+  transaction(usbConnection, [0x9f], 3, null, null, next);
+}
+
+function checkChipId(chipId, next) {
+  if (Buffer.compare(chipId, CHIP_ID) !== 0) {
+    return next(new Error('Invalid chip ID (flash communication error)'));
+  }
+  next();
+}
+
+function setWriteEnabled(usbConnection, next) {
+  transaction(usbConnection, [0x06], null, null, null, next);
+}
+
+function eraseChip(usbConnection, next) {
+  transaction(usbConnection, [0x60], null, null, null, next);
+}
+
+// Poll for the WIP bit in the status register to go low
+function waitTransactionComplete(usbConnection, next) {
+  setTimeout(function onWait() {
+    getStatus(usbConnection, function onGetStatus(err, status) {
+      if (err) {
+        return next(err);
+      }
+      if (status === 0) {
+        return next();
+      }
+      process.nextTick(waitTransactionComplete, usbConnection, next);
+    });
+  }, 200);
+}
+
+// Read the status register
+function getStatus(usbConnection, next) {
+  transaction(usbConnection, [0x05], 1, null, null, function onGetStatus(err, data) {
+    next(err, data ? data[0] : data);
+  });
+}
+
+function write(usbConnection, writeAddress, data, sliceStart, next) {
+  if (!next) {
+    next = sliceStart;
+    sliceStart = 0;
+  }
+  if (sliceStart >= data.length) {
+    return next();
+  }
+
+  var sliceEnd = sliceStart + PAGE;
+  var pageData = data.slice(sliceStart, sliceEnd);
+  writePage(usbConnection, writeAddress, pageData, function onWrite(err) {
+    if (err) {
+      return next(err);
+    }
+    process.nextTick(write, usbConnection, writeAddress + PAGE, data, sliceEnd, next);
+  });
+}
+
+function concatToBuffer() {
+  var totalLength = 0;
+  var buffers = [];
+  var buffer;
+  for (var i = 0; i < arguments.length; i++) {
+    var arg = arguments[i];
+    buffer = new Buffer(arg);
+    totalLength += buffer.length;
+    buffers.push(buffer);
+  }
+  var result = new Buffer(totalLength);
+
+  var byteIndex = 0;
+  for (i = 0; i < buffers.length; i++) {
+    buffer = buffers[i];
+    buffer.copy(result, byteIndex);
+    byteIndex += buffer.length;
+  }
+
+  return result;
+}
```

Comment:
In addition to arrays, there are also buffers coming into this function (from `  var pageData = data.slice(sliceStart, sliceEnd);` in the `write` function) so I went with this:

```
function toBuffer() {
  return Buffer.concat( Array.from(arguments).map (
    (arg) => Array.isArray(arg) ? new Buffer(arg) : arg
  ));
}
```